### PR TITLE
netcdf4 complex number support

### PR DESCRIPTION
**Pull request by piyushrpt**
## What does this PR do?
- Adds support for reading complex datasets from netcdf4
- Works with data generated in a manner recommended by python netCDF4 library - https://github.com/Unidata/netcdf4-python/blob/master/examples/tutorial.py#L133-L149
- Includes same unit tests as the ones for HDF5 driver

## What are related issues/pull requests?
- Adds support in a manner that is similar to the HDF5 driver 

## Tasklist

 - [ ] ADD YOUR TASKS HERE
 - [ ] Add test case(s)
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

### CODE REVIEW

**Comment by rouault** [comment]
With netcdf 4.1.1, 
```
netcdfdataset.cpp:301:24: error: â€˜NC_FIRSTUSERTYPEIDâ€™ was not declared in this scope
     if (nc_datatype >= NC_FIRSTUSERTYPEID)
```

**Comment by piyushrpt** [comment]
https://www.unidata.ucar.edu/support/help/MailArchives/netcdf/msg10845.html

Have the other issues resolved. Will try to find what was done to address this missing variable and implement it.

File: gdal/frmts/netcdf/netcdfdataset.cpp [hunk 1]
```diff
@@ -10963,3 +11119,34 @@ static CPLErr NCDFGetCoordAndBoundVarFullNames( int nCdfId,
...
+    //Adapted from OPENDAP netcdf_handler
+    //To circumvent use of NC_FIRSTUSERTYPEID
+    //Which is not a part of netcdf 4.1.1 installed on RH
+    //In all later version, type >= NC_FIRSTUSERTYPEID works
+#if NETCDF_HAS_NC4
+    int ntypes;
+    int typeids[NC_MAX_VARS];  // It's likely safe to assume there are
+    int rootid; //For nested datasets
+
+    ERR_RET(NCDFGetRootGroup(ncid, &rootid));
```
Comment (rouault) [inline gdal/frmts/netcdf/netcdfdataset.cpp]: c:\projects\gdal\gdal\frmts\netcdf\netcdfdataset.cpp(11135): warning C4800: 'CPLErr': forcing value to bool 'true' or 'false' (performance warning)

ERR_RET returns a CPLErr whereas this method returns bool. You should do the error checking manually here

File: gdal/frmts/netcdf/netcdfdataset.cpp [hunk 2]
```diff
@@ -10963,3 +11119,39 @@ static CPLErr NCDFGetCoordAndBoundVarFullNames( int nCdfId,
...
+    //In all later version, type >= NC_FIRSTUSERTYPEID works
+#if NETCDF_HAS_NC4
+    int ntypes;
+    int typeids[NC_MAX_VARS];  // It's likely safe to assume there are
+    int rootid; //For nested datasets
+
+    if (NCDFGetRootGroup(ncid, &rootid) != CE_None)
+    {
+        CPLError(CE_Failure, CPLE_AppDefined,
+                "Could not get root group information for user defined types");
```
Comment (rouault) [inline gdal/frmts/netcdf/netcdfdataset.cpp]: missing return false;

**Comment by piyushrpt** [comment]
Looks like the failure is unrelated to the netcdf4 updates.

**Pull request merged**